evm_arithmetization/src/generation/prover_input.rs

```diff
@@ -513,103 +517,92 @@ impl<F: RichField> GenerationState<F> {
-            |&(_, [prev_addr, prev_key, ..], [next_addr, next_key, ..])| {
-                let prev_is_less_or_equal = (prev_addr < addr || prev_addr == U256::MAX)
-                    || (prev_addr == addr && prev_key <= key);
-                let next_is_strictly_larger =
-                    next_addr > addr || (next_addr == addr && next_key > key);
-                prev_is_less_or_equal && next_is_strictly_larger
-            },
-        ) {
-            Ok((pred_ptr - U256::from(Segment::StorageLinkedList as usize))
-                / U256::from(STORAGE_LINKED_LIST_NODE_SIZE))
-        } else {
-            Ok(U256::zero())
+
+        let (&(pred_addr, pred_slot_key), &pred_ptr) =
+            self.storage.range(..=(addr, key)).next_back().unwrap_or((
+                &(U256::MAX, U256::zero()),
+                &(Segment::StorageLinkedList as usize),
+            ));
+        if (pred_addr != addr || pred_slot_key != key) && input_fn.0[1] == "insert_slot" {
+            self.storage.insert(
+                (addr, key),
+                Segment::StorageLinkedList as usize
+                    + self
+                        .memory
+                        .get_preinit_memory(Segment::StorageLinkedList)
+                        .len(),
+            );
         }
+        Ok(U256::from(
+            (pred_ptr - Segment::StorageLinkedList as usize) / STORAGE_LINKED_LIST_NODE_SIZE,
+        ))
     }
 
-    /// Returns a pointer `ptr` to a node of the form [next_addr, ..]  in the
+    /// Returns a pointer `ptr` to a node of the form [..] -> [next_addr, ..]
     /// list such that `next_addr = addr` and `addr` is the top of the stack.
     /// If the element is not in the list, loops forever.
-    fn run_next_remove_account(&self) -> Result<U256, ProgramError> {
+    fn run_next_remove_account(&mut self) -> Result<U256, ProgramError> {
         let addr = stack_peek(self, 0)?;
-        let accounts_mem = self.memory.get_preinit_memory(Segment::AccountsLinkedList);
-        let accounts_linked_list =
-            LinkedList::<ACCOUNTS_LINKED_LIST_NODE_SIZE>::from_mem_and_segment(
-                &accounts_mem,
-                Segment::AccountsLinkedList,
-            )?;
-
-        if let Some(([.., ptr], _, _)) = accounts_linked_list
-            .tuple_windows()
-            .find(|&(_, _, [next_node_addr, ..])| next_node_addr == addr)
-        {
-            Ok(ptr / ACCOUNTS_LINKED_LIST_NODE_SIZE)
-        } else {
-            Ok((Segment::AccountsLinkedList as usize).into())
-        }
+
+        let (_, &ptr) = self
+            .accounts
+            .range(..addr)
+            .next_back()
+            .unwrap_or((&U256::MAX, &(Segment::AccountsLinkedList as usize)));
+        self.accounts
+            .remove(&addr)
+            .ok_or(ProgramError::ProverInputError(InvalidInput))?;
+
+        Ok(U256::from(ptr / ACCOUNTS_LINKED_LIST_NODE_SIZE))
```

Comment:
Same here, we'll be returning `Segment::AccountsLinkedList / ACCOUNTS_LINKED_LIST_NODE_SIZE` in case it's not found, instead of `Segment::AccountsLinkedList`

Reply:
In case not found it will return an error because of
```rust
self.accounts
            .remove(&addr)
            .ok_or(ProgramError::ProverInputError(InvalidInput))?;
``` 
`Segment::AccountsLinkedList / ACCOUNTS_LINKED_LIST_NODE_SIZE` is returned when you're searching the minimum (and hence the special node is the predecesor)
The original code would loop forever, which would cause the else branch to never be executed